File: october_2024/potd_19_10_2024.cpp
//Nearest multiple of 10


class Solution {
public:
string roundToNearest(string str) {
        // Complete the function
        int n = str.length();
        
        // Check the last digit to determine rounding
        if (str[n-1] <= '5') {
            // Round down by setting the last digit to '0'
            str[n-1] = '0';
            return str;
        }
        // Round up
        str[n-1] = '0';  // Set last digit to '0'
        int i = n - 2;    // Start from the second last digit
        
        // Handle carry-over in case of '9'
        while (i >= 0 && str[i] == '9') {
            str[i] = '0'; // Set all '9's to '0'
            i--;
        }

        // If all digits were '9', prepend '1'
        if (i < 0) {
            str = '1' + str;
        } else {
            // Increase the current digit
            str[i]++;
        }

        return str;
    }
};

File: october_2024/potd_19_10_2024.java
// https://www.geeksforgeeks.org/problems/nearest-multiple-of-102437/1

class Solution {
    String roundToNearest(String str) {
        char[] numArr = str.toCharArray();
        int n = numArr.length;

        if (numArr[n - 1] <= '5') {
            numArr[n - 1] = '0';
        } else {
            numArr[n - 1] = '0';
            int i = n - 2;
            boolean flag = false;

            while (i >= 0) {
                if (numArr[i] == '9') {
                    numArr[i] = '0';
                } else {
                    numArr[i]++;
                    flag = true;
                    break;
                }
                i--;
            }

            if (!flag) {
                return '1' + new String(numArr);
            }
        }

        return new String(numArr);
    }
}